staging/src/k8s.io/client-go/tools/pager/pager.go

```diff
@@ -115,3 +127,86 @@ func (p *ListPager) List(ctx context.Context, options metav1.ListOptions) (runti
 		options.Continue = m.GetContinue()
 	}
 }
+
+// EachListItem invokes fn on each runtime.Object in the list. Any error immediately terminates the
+// loop. The items are retrieved via List in small chunks from the server to reduce the impact on
+// the server. Up to ListPager.PageBufferSize chunks are buffered concurrently in the background.
+// If the chunk attempt fails a "Expired" error may be returned.
+func (p *ListPager) EachListItem(ctx context.Context, options metav1.ListOptions, fn func(obj runtime.Object) error) error {
+	return p.eachListChunkBuffered(ctx, options, func(obj runtime.Object) error {
+		return meta.EachListItem(obj, fn)
+	})
+}
+
+// stoppedErr is an internal error for signaling eachListChunk should stop processing chunks.
+// It must be handled internally and never be returned to API callers.
+var stoppedErr = goerrors.New("paginated list stopped")
+
+// eachListChunkBuffered invokes fn on each runtimeObject list chunk. It buffers up to
+// PageBufferSize chunks of list items concurrently in the background. If the chunk attempt fails a "Expired"
+// error may be returned.
+func (p *ListPager) eachListChunkBuffered(ctx context.Context, options metav1.ListOptions, fn func(obj runtime.Object) error) error {
+	chunkC := make(chan runtime.Object, p.PageBufferSize)
+
+	stopC := make(chan struct{})
+	bgResultC := make(chan error, 1)
+	go func() {
+		err := p.eachListChunk(ctx, options, func(chunk runtime.Object) error {
+			select {
+			case chunkC <- chunk: // buffer the chunk
+			case <-stopC:
+				return stoppedErr
+			}
+			return nil
+		})
+		if err == stoppedErr {
+			err = nil // stoppedErr is an internal signal
```

Comment:
I understand what this line does, returning due to been signaled to stop is not actually an error.

Do we need to define an internal error? Can we use context cancel function to signal `eachListChunk()` to stop? Upon closing `stopC`, `fn` can return nil (line 202), but the loop inside `eachListChunk()` will be stopped by `ctx.Done()`.

Reply:
You're right. A nested context works here. Since the context error should be return iff the caller cancels the context, reasoning through when/how cancel error get's handled is a bit subtle and I've added a comment explaining it.  I've also added test coverage for cancelation, both for then a `fn` error results in an early exit and when the caller calls cancel.